test/new-e2e/tests/windows/common/powershell/command_builder.go

```diff
@@ -0,0 +1,157 @@
+	return ps
+}
+
+// GetActiveDirectoryDomainController creates a command that returns information about the current Active Directory Domain Controller.
+func (ps *powerShellCommandBuilder) GetActiveDirectoryDomainController() *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, "Get-ADDomainController")
+	return ps
+}
+
+// Reboot creates a command that reboots the machine.
+func (ps *powerShellCommandBuilder) Reboot() *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, "Restart-Computer -Force")
+	return ps
+}
+
+// InstallADDSForest creates a command that promotes a server to the role of forest.
+func (ps *powerShellCommandBuilder) InstallADDSForest(activeDirectoryDomain, passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+    CreateDNSDelegation           = $false
+    ForestMode                    = "Win2012R2"
+    DomainMode                    = "Win2012R2"
+    DomainName                    = "%s"
+    SafeModeAdministratorPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+    Force                         = $true
+}; Install-ADDSForest @HashArguments`, activeDirectoryDomain, passwd))
+	return ps
+}
+
+// UninstallADDSDomainController creates a command to remove a Domain Controller.
+func (ps *powerShellCommandBuilder) UninstallADDSDomainController(passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+	SkipPreChecks              = $true
+    LocalAdministratorPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+    DemoteOperationMasterRole  = $true
+	ForceRemoval               = $true
+	Force                      = $true
+}; Uninstall-ADDSDomainController
+`, passwd))
+	return ps
+}
+
+// AddActiveDirectoryUser creates a command for adding a user to an Active Directory domain.
+func (ps *powerShellCommandBuilder) AddActiveDirectoryUser(username, passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+	Name = '%s'
+	AccountPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+	Enabled = $true
+}; New-ADUser @HashArguments
+`, username, passwd))
+	return ps
+}
+
+// GetMachineType creates a command that returns the ProductType of the machine (2 for server, 3 for domain controller).
+func (ps *powerShellCommandBuilder) GetMachineType() *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, "(Get-CimInstance -ClassName Win32_OperatingSystem).ProductType")
+	return ps
+}
+
+// StartService creates a command that starts a given service.
+func (ps *powerShellCommandBuilder) StartService(serviceName string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+Start-Service %s
+`, serviceName))
+	return ps
+}
+
+// WaitForServiceStatus creates a command that waits 1 minute for a service to reach a given state.
+func (ps *powerShellCommandBuilder) WaitForServiceStatus(serviceName, status string) *powerShellCommandBuilder {
```

Comment:
One pain point of .NET's `WaitForStatus` is it only waits for the single status being passed in, so in the failure case you may end up waiting longer than you need to.
For example, when you start a service and wait for it to be `Running`, if the service fails to start its state becomes `Stopped`. There's no point in continuing to wait at this point but `WaitForStatus` will still block until the timeout is met.

I think it's worth adding a comment to that effect.